File: exercise/python/04_workout_tracker.py
"""
Exercise 4: Personal Workout Tracker

You've learned the building blocks: variables, lists, dictionaries, and loops.
Now it's time to organize your code with functions! Functions are like creating your
own custom exercises - you define them once, then use them whenever you need.

Concepts covered:
- Functions (defining reusable blocks of code)
- Function parameters (inputs)
- Return values (outputs)
- Function documentation
- Organizing code with functions

Learning Goals:
- Create functions to organize your code
- Pass data into functions using parameters
- Return calculated results from functions
- Build a workout tracking system using multiple functions

Run this exercise: uv run exercise/python/04_workout_tracker.py
"""

# ============================================================================
# QUICK INTRO: What are functions?
# ============================================================================
# A function is a named block of code that performs a specific task.
# Think of it like a gym routine - you define the routine once, then
# execute it whenever you want.
#
# Define a function:
#   def function_name(parameters):
#       # code here
#       return result
#
# Call a function:
#   result = function_name(arguments)


# ============================================================================
# EXAMPLE CODE
# ============================================================================
def calculate_volume(reps, weight):
    """
    Calculate the volume for a single set.

    Parameters:
        reps: Number of repetitions
        weight: Weight used in pounds

    Returns:
        The volume (reps × weight)
    """
    weight = 50
    volume = reps * weight
    return volume


def calculate_total_volume(reps_list, weights_list):
    """
    Calculate total volume across multiple sets.

    Parameters:
        reps_list: List of reps for each set
        weights_list: List of weights for each set

    Returns:
        Total volume across all sets
    """
    total = 0
    for i in range(len(reps_list)):
        set_volume = calculate_volume(reps_list[i], weights_list[i])
        total += set_volume
    return total


def display_workout_summary(exercise_name, reps_list, weights_list):
    """
    Display a formatted summary of a workout.

    This function doesn't return anything - it just prints output.
    """
    print(f"\n=== {exercise_name} Summary ===")
    print(f"Sets completed: {len(reps_list)}")
    print(f"Total reps: {sum(reps_list)}")

    total_vol = calculate_total_volume(reps_list, weights_list)
    print(f"Total volume: {total_vol} lbs")

    print("\nSet breakdown:")
    for i in range(len(reps_list)):
        set_num = i + 1
        print(f"  Set {set_num}: {reps_list[i]} reps × {weights_list[i]} lbs")


# Test the functions
print("=== Testing Functions ===")

# Call calculate_volume
set_volume = calculate_volume(10, 135)
print(f"Single set volume: {set_volume} lbs")

# Call calculate_total_volume
bench_reps = [10, 8, 8, 6]
bench_weights = [135, 155, 175, 185]
total = calculate_total_volume(bench_reps, bench_weights)
print(f"Total bench press volume: {total} lbs")

# Call display_workout_summary
display_workout_summary("Bench Press", bench_reps, bench_weights)


# ============================================================================
# YOUR TURN: TODO EXERCISES
# ============================================================================

print("\n\n=== YOUR WORKOUT TRACKER ===")

# TODO 1: Create a function to calculate one-rep max (1RM)
# The Brzycki formula for estimating 1RM is: weight × (36 / (37 - reps))
#
# Create a function called calculate_one_rep_max that takes two parameters:
# - weight: the weight lifted
# - reps: number of reps performed
#
# The function should return the estimated 1RM
#
# HINT: Use the formula above and return the result
#
# Stuck? Ask: "How do I create a function with parameters in Python?"
# Or: "What does the return keyword do?"

# Write your code here:

# TODO 2: Test your one_rep_max function
# Call your calculate_one_rep_max function with weight=225 and reps=5
# Store the result in a variable called estimated_max
# Print the result with a descriptive message
#
# Need help? Ask: "How do I call a function and use its return value?"

# Write your code here:


# TODO 3: Create a function to log a workout
# Create a function called log_workout that takes three parameters:
# - exercise_name: name of the exercise
# - sets: number of sets
# - reps: number of reps per set
#
# The function should print a message like:
# "Logged: 4 sets of 8 reps of Squats"
#
# This function doesn't need to return anything (it just prints)
#
# HINT: Use print() with an f-string inside the function
#
# Stuck? Ask: "How do I create a function that prints but doesn't return anything?"

# Write your code here:





# TODO 4: Test your log_workout function
# Call log_workout with:
# - exercise_name: "Deadlift"
# - sets: 5
# - reps: 5
#
# Need help? Ask: "How do I call a function with multiple arguments?"

# Write your code here:




# TODO 5: Create a function to calculate workout duration
# Create a function called calculate_workout_duration that takes:
# - num_sets: number of sets
# - rest_time: rest time between sets in seconds
# - work_time_per_set: time spent working per set in seconds
#
# Calculate total time: (num_sets × work_time_per_set) + ((num_sets - 1) × rest_time)
# Note: There's rest between sets, but not after the last set
#
# Return the total time in minutes (divide by 60)
#
# HINT: Do the calculation, then divide by 60 before returning
#
# Stuck? Ask: "How do I calculate total workout time with rest periods?"

# Write your code here:


# TODO 6: Test your workout duration function
# Call calculate_workout_duration with:
# - 5 sets
# - 90 seconds rest
# - 30 seconds work time per set
#
# Store the result and print how long the workout will take
#
# Need help? Ask: "Show me an example of calling a function with three parameters"

# Write your code here:




# ============================================================================
# BONUS CHALLENGES (Optional)
# ============================================================================

# BONUS 1: Create a workout comparison function
# Create a function called compare_workouts that takes two volume values
# and returns which one is higher (or if they're equal)
#
# Parameters: volume1, volume2
# Return: A string like "Workout 1 had higher volume" or "Equal volume"
#
# For help: Ask "How do I compare two values and return different strings?"

# Write your code here:




# BONUS 2: Create a comprehensive workout report function
# Create a function called generate_workout_report that takes:
# - exercise_name, reps_list, weights_list
#
# The function should:
# 1. Calculate and display total volume
# 2. Calculate and display average reps
# 3. Find and display the heaviest weight used
# 4. Estimate the 1RM from the first set
#
# Want guidance? Ask: "How do I create a function that does multiple calculations?"

# Write your code here:




# BONUS 3: Create a rest recommendation function
# Create a function called recommend_rest_time that takes a weight parameter
# and returns recommended rest time in seconds:
# - Light weight (< 135 lbs): 60 seconds
# - Medium weight (135-225 lbs): 90 seconds
# - Heavy weight (> 225 lbs): 180 seconds
#
# Need help? Ask: "How do I use if/elif/else in a function?"

# Write your code here:




# ============================================================================
# TESTING YOUR CODE
# ============================================================================

# Test cases for your functions:
#
# calculate_one_rep_max(225, 5) should return approximately 258.75
# log_workout("Deadlift", 5, 5) should print "Logged: 5 sets of 5 reps of Deadlift"
# calculate_workout_duration(5, 90, 30) should return approximately 8.5 minutes
#
# If your functions aren't working as expected:
# - Check that you're using the correct parameter names
# - Make sure you're returning values (not just printing them) when needed
# - Verify your math in the calculations
#
# Ask me: "Can you help me debug my function?"


if __name__ == "__main__":
    # You can add test calls here to run when the file executes
    pass

File: exercise/typescript/04_workout_tracker.ts
/**
 * Exercise 4: Personal Workout Tracker with Typed Functions
 *
 * You've mastered types, arrays, and interfaces. Now let's organize code with FUNCTIONS!
 * In TypeScript, functions have types too - for parameters AND return values.
 * This makes your functions self-documenting and catches errors when you pass wrong data.
 *
 * IMPORTANT: Function parameters MUST have explicit types (inference doesn't work here).
 * Return types CAN be inferred, but explicit types make code clearer.
 *
 * Concepts covered:
 * - Function parameter types (required!)
 * - Function return types
 * - Void return type (for functions that don't return)
 * - Optional parameters
 * - Function type annotations
 * - Type safety in function calls
 *
 * Run this exercise: node exercise/typescript/04_workout_tracker.ts
 */

// ============================================================================
// QUICK INTRO: Typed Functions
// ============================================================================
// In TypeScript, functions need types for parameters and (optionally) return values:
//
//   function add(a: number, b: number): number {
//     return a + b;
//   }
//
// - Parameters: MUST have explicit types (a: number, b: number)
// - Return type: Optional but recommended (: number)
//
// TypeScript checks:
// - You pass the right types when calling
// - The function returns the declared type
// - All code paths return a value (if return type is not void)

// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

// We can define the type of a function itself!
type VolumeCalculator = (reps: number, weight: number) => number;

// ============================================================================
// EXAMPLE CODE
// ============================================================================

/**
 * Calculate volume for a single set.
 *
 * @param reps - Number of repetitions
 * @param weight - Weight in pounds
 * @returns The volume (reps × weight)
 */
function calculateVolume(reps: number, weight: number): number {
    const volume = reps * weight;
    return volume;
}

/**
 * Calculate total volume across multiple sets.
 */
function calculateTotalVolume(reps: number[], weights: number[]): number {
    let total = 0;
    for (let i = 0; i < reps.length; i++) {
        total += calculateVolume(reps[i], weights[i]);
    }
    return total;
}

/**
 * Display a workout summary (no return value).
 * The void return type means this function doesn't return anything.
 */
function displayWorkoutSummary(
    exerciseName: string,
    reps: number[],
    weights: number[]
): void {
    
    console.log(`\n=== ${exerciseName} Summary ===`);
    console.log(`Sets completed: ${reps.length}`);
    console.log(`Total reps: ${reps.reduce((sum, r) => sum + r, 0)}`);

    const totalVol = calculateTotalVolume(reps, weights);
    console.log(`Total volume: ${totalVol} lbs`);

    console.log("\nSet breakdown:");
    for (let i = 0; i < reps.length; i++) {
        console.log(`  Set ${i + 1}: ${reps[i]} reps × ${weights[i]} lbs`);
    }
}

/**
 * Function with optional parameter (uses ?)
 */
function logWorkout(exercise: string, sets: number, reps?: number): void {
    if (reps !== undefined) {
        console.log(`Logged: ${sets} sets of ${reps} reps of ${exercise}`);
    } else {
        console.log(`Logged: ${sets} sets of ${exercise}`);
    }
}

// Test the functions
console.log("=== Testing Functions ===");

// TypeScript checks the parameter types!
const setVolume = calculateVolume(10, 135);
console.log(`Single set volume: ${setVolume} lbs`);

// Try uncommenting this to see a type error:
// calculateVolume("10", 135); // ERROR! String is not assignable to number

const benchReps = [10, 8, 8, 6];
const benchWeights = [135, 155, 175, 185];
const total = calculateTotalVolume(benchReps, benchWeights);
console.log(`Total volume: ${total} lbs`);

displayWorkoutSummary("Bench Press", benchReps, benchWeights);

console.log();
logWorkout("Squats", 5, 5); // With reps
logWorkout("Plank", 3);      // Without reps (optional)

// ============================================================================
// YOUR TURN: TODO EXERCISES
// ============================================================================

console.log("\n\n=== YOUR WORKOUT TRACKER ===");

// TODO 1: Create a function to calculate one-rep max
// The Brzycki formula: weight × (36 / (37 - reps))
//
// Create a function called calculateOneRepMax with:
// - Parameters: weight (number), reps (number)
// - Return type: number
// - Calculate and return the estimated 1RM
//
// Type hint: function name(param: type, param: type): returnType { }
//
// Stuck? Ask: "How do I create a function with typed parameters in TypeScript?"

// Write your code here:






// TODO 2: Test your calculateOneRepMax function
// Call your function with weight=225, reps=5
// Store the result in a variable with explicit type: number
// Print the result with a message
//
// Type hint: const result: number = functionName(arg1, arg2);
//
// Need help? Ask: "How do I call a typed function in TypeScript?"

// Write your code here:





// TODO 3: Create a function with void return type
// Create a function called logWorkoutSession that:
// - Parameters: exerciseName (string), sets (number), reps (number)
// - Return type: void (doesn't return anything, just prints)
// - Prints: "Logged: X sets of Y reps of ExerciseName"
//
// Type hint: Functions that only print should return void
//
// Stuck? Ask: "What does void mean in TypeScript?"

// Write your code here:






// TODO 4: Test your logWorkoutSession function
// Call it with: "Deadlift", 5, 5
// Then call it again with different values
//
// Type challenge: Try passing wrong types and see TypeScript catch it!
//
// Need help? Ask: "How do I call a function that returns void?"

// Write your code here:





// TODO 5: Create a function to calculate workout duration
// Create a function called calculateWorkoutDuration with:
// - Parameters: numSets (number), restTime (number), workTimePerSet (number)
// - Return type: number
// - Calculate: (numSets × workTimePerSet) + ((numSets - 1) × restTime)
// - Return the result in MINUTES (divide by 60)
//
// Type hint: There's rest between sets but not after the last set
//
// Stuck? Ask: "How do I create a function that does calculations and returns a value?"

// Write your code here:







// TODO 6: Test workout duration function with explicit typing
// Call calculateWorkoutDuration with: 5 sets, 90 seconds rest, 30 seconds work
// Store result in a variable with explicit number type
// Print how long the workout will take
//
// Type hint: const duration: number = ...
//
// Need help? Ask: "How do I store and use a function's return value?"

// Write your code here:





// TODO 7: Create a function with optional parameters
// Create a function called trackWorkout with:
// - Required params: exerciseName (string), sets (number)
// - Optional param: notes (string) - use ? to make it optional
// - Return type: void
// - If notes provided, print: "ExerciseName: X sets - Notes"
// - If no notes, print: "ExerciseName: X sets"
//
// HINT: Use ? after parameter name to make it optional
// HINT: Check if optional param exists: if (notes !== undefined) { }
//
// Stuck? Ask: "How do I create a function with optional parameters in TypeScript?"

// Write your code here:







// TODO 8: Test the optional parameter function
// Call trackWorkout twice:
// 1. With notes: "Bench Press", 3, "Felt strong today"
// 2. Without notes: "Squats", 4
//
// Type hint: Optional parameters can be omitted in the call
//
// Need help? Ask: "How do I call a function with optional parameters?"

// Write your code here:





// ============================================================================
// BONUS CHALLENGES (Optional)
// ============================================================================

// BONUS 1: Create a function that returns an object
// Create a function called analyzeWorkout that takes:
// - reps: number[]
// - weights: number[]
// Returns an object with:
// - totalReps: number
// - totalVolume: number
// - averageWeight: number
// - maxWeight: number
//
// Type hint: You can specify return type as an object:
// function name(): { prop1: type, prop2: type } { }
// OR create an interface for the return type!
//
// For help: Ask "How do I create a function that returns an object in TypeScript?"

// Write your code here:







// BONUS 2: Create a function with a union type parameter
// Create a function called formatWeight that takes:
// - weight: number | string (can be either!)
// - Returns: string
// If weight is a number, return: "X lbs"
// If weight is a string, return: weight (as-is)
//
// HINT: Use typeof to check type: if (typeof weight === "number")
//
// Want guidance? Ask: "How do I use union types in function parameters?"

// Write your code here:







// BONUS 3: Create a higher-order function
// Create a function called createVolumeCalculator that:
// - Takes a multiplier: number
// - Returns a FUNCTION that calculates volume with that multiplier
// - The returned function signature: (reps: number, weight: number) => number
//
// This is advanced! Functions can return functions!
//
// HINT:
// function createCalculator(mult: number): (r: number, w: number) => number {
//   return (reps, weight) => reps * weight * mult;
// }
//
// Need help? Ask: "What are higher-order functions in TypeScript?"

// Write your code here:







// ============================================================================
// TESTING YOUR CODE
// ============================================================================

// Expected behavior:
// - calculateOneRepMax(225, 5) ≈ 258.75
// - logWorkoutSession prints formatted workout logs
// - calculateWorkoutDuration(5, 90, 30) ≈ 8.5 minutes
// - trackWorkout works with and without optional notes parameter
//
// TypeScript ensures:
// - You can't pass wrong types to functions
// - Functions return the declared type
// - Required parameters are always provided
// - Optional parameters work correctly
//
// Key learning points:
// - Parameter types are REQUIRED (can't be inferred)
// - Return types are optional but recommended for clarity
// - void means "no return value"
// - Optional parameters use ? and can be omitted
// - TypeScript checks function calls for type correctness
//
// If you get type errors:
// - Check parameter types match when calling
// - Verify return type matches what you're returning
// - Ensure all code paths return a value (unless void)
// - Check optional parameter usage
//
// Ask me: "Why is TypeScript showing an error in my function?"
